sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java

```diff
@@ -0,0 +1,398 @@
+          .apply("Identity", ParDo.of(new DoFn<KV<String, String>, KV<String, String>>() {
+            @ProcessElement
+            public void processElement(ProcessContext context) {
+              context.output(context.element());
+            }
+      }).withSideInputs(empty));
+      return materialized
+          .apply("Pair with random key", ParDo.of(
+              new DoFn<KV<String, String>, KV<Integer, KV<String, String>>>() {
+                private int shard;
+
+                @Setup
+                public void setup() {
+                  shard = ThreadLocalRandom.current().nextInt();
+                }
+
+                @ProcessElement
+                public void processElement(ProcessContext context) {
+                  context.output(KV.of(++shard, context.element()));
+                }
+              }
+          ))
+          .apply(Reshuffle.<Integer, KV<String, String>>of())
+          .apply(Values.<KV<String, String>>create());
+    }
+  }
+
+  /**
+   * A {@link PTransform} to write to a Redis server.
+   */
+  @AutoValue
+  public abstract static class Write extends PTransform<PCollection<KV<String, String>>, PDone> {
+
+    @Nullable abstract RedisConnectionConfiguration connectionConfiguration();
+
+    abstract Builder builder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+
+      abstract Builder setConnectionConfiguration(
+          RedisConnectionConfiguration connectionConfiguration);
+
+      abstract Write build();
+
+    }
+
+    public Write withConnectionConfiguration(RedisConnectionConfiguration connection) {
+      checkArgument(connection != null, "connection can not be null");
+      return builder().setConnectionConfiguration(connection).build();
+    }
+
+    @Override
+    public PDone expand(PCollection<KV<String, String>> input) {
+      checkArgument(connectionConfiguration() != null, "withConnectionConfiguration() is required");
+
+      input.apply(ParDo.of(new WriteFn(this)));
+      return PDone.in(input.getPipeline());
+    }
+
+    private static class WriteFn extends DoFn<KV<String, String>, Void> {
+
+      private static final int DEFAULT_BATCH_SIZE = 1000;
+
+      private final Write spec;
+
+      private transient Jedis jedis;
+      private transient Pipeline pipeline;
+
+      private int batchCount;
+
+      public WriteFn(Write spec) {
+        this.spec = spec;
+      }
+
+      @Setup
+      public void setup() {
+        jedis = spec.connectionConfiguration().connect();
+      }
+
+      @StartBundle
+      public void startBundle() {
+        pipeline = jedis.pipelined();
+        pipeline.multi();
+        batchCount = 0;
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext processContext) {
+        KV<String, String> record = processContext.element();
+        pipeline.append(record.getKey(), record.getValue());
+
+        if (batchCount >= DEFAULT_BATCH_SIZE) {
```

Comment:
batchCount is not incremented anywhere, nor reset back to 0 in flush()

Reply:
Good catch, thanks !